pkg/resourcecollector/virtualmachine.go

```diff
@@ -5,6 +5,102 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 )
 
+/*
+ * transform dataVolume: to persistentVolumeClaim:
+ *	       name:	      claimName:
+ */
+func transformDataVolume(input map[string]interface{}, name string ) map[string]interface{} {
+
+	output := make(map[string]interface{})
+
+	for key, val := range input {
+		if key == name {
+			output["claimName"] = val
+			continue
+		}
+		output[key] = val
+	}
+	return output
+}
+
+/*
+ * Utility function for virtualMachine template parsing logic
+ */
+func parseMap(input map[string]interface{}, output map[string]interface{}, key string, is_key_present *bool) {
+
+	for mKey, mVal := range input {
+		switch mKey {
+		case "macAddress":
+			*is_key_present = true
+			continue
+		case "dataVolume":
+			*is_key_present = true
+			switch typeval := mVal.(type) {
+			case map[string]interface{}:
+				output["persistentVolumeClaim"] =  transformDataVolume(typeval, "name")
+				continue
+			}
+		}
+		output[mKey] = mVal
+	}
+}
+
+/*
+ * Path parsing function for virtualMachine template.
+ */
+func transformPath(object runtime.Unstructured, path []string, key string) error {
+
+        content := object.UnstructuredContent()
+        data, _, er := unstructured.NestedSlice(content, path...)
+        if er != nil {
+		return er
+        }
+        found := false
+        interf := make([]interface{}, 0, 1)
+        for _, val := range data {
+		switch typeval := val.(type) {
+		case map[string]interface{}:
+			newMap := make(map[string]interface{})
+			parseMap(typeval, newMap, key, &found)
+			interf = append(interf, newMap)
+		}
+        }
+	// Update only if the desired path,value is found in the template for transformation
+	// Ignore otherwise.
+        if found == true {
+		unstructured.RemoveNestedField(content, path...)
+		if err := unstructured.SetNestedSlice(content, interf, path...); err != nil {
+			return err
+		}
+        }
+        return nil
+
+}
+
+/*
+ * This should be called from restore workflow for any resource transform for
+ * VM resources.
+ */
+func (r *ResourceCollector) prepareVirtualMachineForApply(
+	object runtime.Unstructured,
+	namespaces []string,
+) error {
+
+	// Remove macAddress as it creates macAddress conflict while restore
+	// to the same cluster to different namespace while the source VM is up and
+	path := []string{"spec", "template", "spec", "domain", "devices", "interfaces"}
```

Comment:
Can we have the path as `.` separated for easier understanding. eg `spec.template.spec.domain.devices.interfaces`